### PR TITLE
refactor: store test's metadata directly in test

### PR DESCRIPTION
**Pull request by kpodsiad**
Javascript is a dynamic language so it's possible to extend object and classes with new fields.

### CODE REVIEW

**Review by kpodsiad — COMMENTED** [review]
As I said, Javascript is dynamic and it allows to add the new fields to an existing object. I used this property to get rid off creating and keeping detached metadata per every test item. Instead, this information can be stored directly in the test item object.

In my opinion, it's a better approach and it reduces the number of unnecessary ifs (null & undefined guarding)

File: src/testExplorer/util.ts [hunk 1]
```diff
@@ -1,5 +1,41 @@
...
+} from "./types";
+
+// prettier-ignore
+export function refineTestItem(test: vscode.TestItem, kind: "project", targetUri: TargetUri, targetName: TargetName): ProjectMetalsTestItem;
+// prettier-ignore
+export function refineTestItem(test: vscode.TestItem, kind: "package", targetUri: TargetUri, targetName: TargetName, parent: vscode.TestItem): PackageMetalsTestItem;
+// prettier-ignore
+export function refineTestItem(test: vscode.TestItem, kind: "suite", targetUri: TargetUri, targetName: TargetName, parent: vscode.TestItem): SuiteMetalsTestItem;
+// prettier-ignore
+export function refineTestItem(test: vscode.TestItem, kind: "testcase", targetUri: TargetUri, targetName: TargetName, parent: vscode.TestItem): TestCaseMetalsTestItem;
```
[the remark this answers is not in the document]
Comment (kpodsiad) [reply src/testExplorer/util.ts:21]: so in other words, you cannot call `refine` without parent if passed kind indicates that parent has to be present

File: src/testExplorer/util.ts [hunk 2]
```diff
@@ -1,5 +1,41 @@
...
+// prettier-ignore
+export function refineTestItem(
+  test: vscode.TestItem,
+  kind: MetalsTestItemKind,
+  targetUri: TargetUri,
+  targetName: TargetName,
+  parent?: vscode.TestItem
+): MetalsTestItem {
+  const casted = test as MetalsTestItem;
```
Comment (tgodzik) [inline src/testExplorer/util.ts:30]:
```suggestion
  const cast = test as MetalsTestItem;
```
I think it's `cast -> cast -> cast`

File: src/testExplorer/util.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply src/testExplorer/util.ts:21]: Shouldn't this be an actual class with a constructor? It could all have `def toMetalsItem`

**Review by tgodzik — CHANGES_REQUESTED** [review]
Great work! I have some minor comments.

**Comment by kpodsiad** [comment]
@tgodzik I added some comments to the overloaded function. Do you still find it confusing?

File: src/testExplorer/util.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply src/testExplorer/util.ts]: Ach, I think I understand what is happening now :S and I feel it's pretty evil :sweat_smile:

**Review by tgodzik — APPROVED** [review]
LGTM

**Pull request merged**